cpp/src/datetime/datetime_ops.cu

```diff
@@ -88,32 +90,60 @@ struct extract_component_operator {
   }
 };
 
-template <datetime_component COMPONENT>
-struct ceil_timestamp {
+// This functor takes the rounding type as runtime info and dispatches to the ceil/floor/round
+// function.
+template <typename DurationType>
+struct RoundFunctor {
+  template <typename Timestamp>
+  CUDA_DEVICE_CALLABLE auto operator()(rounding_kind round_kind, Timestamp dt)
+  {
+    switch (round_kind) {
+      case rounding_kind::CEIL: return cuda::std::chrono::ceil<DurationType>(dt);
+      case rounding_kind::FLOOR: return cuda::std::chrono::floor<DurationType>(dt);
+      case rounding_kind::ROUND: return cuda::std::chrono::round<DurationType>(dt);
+      default: cudf_assert(false && "Unsupported rounding kind.");
+    }
+    __builtin_unreachable();
+  }
+};
+
+struct RoundingDispatcher {
+  rounding_kind round_kind;
+  datetime_component component;
+
```

Comment:
@vyasr is there an example of this in the file? looking at `datetime_ops.cu`, I see the same format we followed.

Reply:
Odd. libcudf typically follows the opposite convention because most of the time members are not part of the public namespace, but of course with structs they are exposed nonetheless. Most style guides that I've seen recommend putting the public API declarations at the top since that's what a reader/user needs. Let's leave this as is for now.